# Что такое функции 

### Обычно функция определяется с помощью инструкции def.
Пример

In [1]:

def func(): 
        
    pass # заполнении функции , ничего не делает 




func()
print( func() )


None


In [4]:

def name():
        
    text=1

    print(text)


name()


1


# Каждая функция может что то возвращать при помощи return , при помощи него функция может сразу заканчиваться , если retrun нет , то она будет идти последовательно пока не выполнит все операции, а если есть return , то сразу закончится и возвратит то значение которое мы передали в return 
Пример

In [16]:


def dima():
    a=2
    b=0
    c=a-b
    return c

def name1():
    a=1
    if a==0:
        return 1
    else:

        a = 2
        b=1
        c=a+b
        return c

#print(dima())

func= name1()
#print(func)
new_1= func + 1000
print(func)
print(new_1)
# name1()
#print(name1())

# print(name1()) # вызов функции с print 

#name1()



3
1003


In [24]:

def sum(as12, qwe):

    first= as12
    second= qwe

    return (as12+second)

a = int(input('Enter 1st number: '))
b = int(input('Enter 2nd number: '))

#print(   f'Sum of {a} and {b} is {sum(a, b)}'   )
print( 'sum of ', a, 'and', b, 'is', sum(a,b))


Sum of 5 and 10000 is 10005


### Функция может принимать значения 

In [22]:
def func(x,y, q=77): # q - необязательная переменная 
    print(x)
    print(y)
    print(q)
    return ( x+y+q )



#print(func(57,999))


print(   func(1,99,10000)      )

1
99
10000
10100


In [23]:
def func(x,y):

    return len(x+y)



print(func('qwe','ewq'))

6


### Может быть вложенной или через себя вызывать другую функцию , тогда это будет стек (проходили его в первом модуле)
### Рекурсивная функция 

In [ ]:

def newfunc(n):

     def myfunc(x):
         return x + n # относится к  myfunc

     return myfunc # относится к newfunc

new = newfunc(100) # # new - это функция
print(new)
print(new(200)) 


In [30]:
def func(a, b, c=2): # c - необязательный аргумент

     return a + b + c



print(func(1, 2) ) # a = 1, b = 2, c = 2 (по умолчанию)

print(func(1, 2, 3) ) # a = 1, b = 2, c = 3
print(func(a=1, b=3) ) # a = 1, b = 3, c = 2
print(func(1 ,1111, c=3) ) # a = 3, c = 6, b не определен

5
6
6
1115


### Функция также может принимать переменное количество позиционных аргументов, тогда перед именем ставится *:

In [ ]:
def func(*args): 

    return args


#print(   func(1, 2, 3, 'abc') )

#print( func() )

func(1)


### Как видно из примера, args - это кортеж из всех переданных аргументов функции, и с переменной можно работать также, как и с кортежем.

### Функция может принимать и произвольное число именованных аргументов, тогда перед именем ставится **
### В переменной kwargs у нас хранится словарь, с которым мы, опять-таки, можем делать все, что нам заблагорассудится.

In [ ]:
def func(**kwargs):

    return kwargs


#print(func(a=1, b=2, c=3))
#func()
func(a='python')

In [1]:
a = 1
a='2222,hi'
a=1.5555
a=True
a= False 

a=[1,2]
a.append(2)
b=len(a)


# Рекурсивные функции
Нахождение факториала  (!!!!!!)

In [3]:

def fac( n ):
    
    if n == 0:
        return 1

    return fac(n-1) * n  
 


print(fac(5))

120


0 шаг. Вызов функции: fac(5)
1. fac(5) возвращает fac(4) * 5
2. fac(4) => fac(3) * 4
3. fac(3) => fac(2) * 3
4. fac(2) => fac(1) * 2
5. fac(1) => 1
6. 1 * 2 - возврат в вызов fac(2)
7. 2 * 3 - fac(3)
8. 6 * 4 - fac(4)
9. 24 * 5 – fac(5)
10. Возврат в основную ветку программы значения 120.

# Анонимные функции, инструкция lambda

In [1]:
func = lambda x, y: x + y

func(1, 2)

3

In [ ]:
(lambda x, y: x + y)(1, 2)
#(lambda x, y: x + y)('a', 'b')

# Что такое map()

In [ ]:
def square(number):
       return number ** 2


numbers = [1, 2, 3, 4, 5]
squared = map(square, numbers)

list(squared)
